### PR TITLE
Support full GitHub app authentication

### PR DESCRIPTION
**Pull request by dblanchette**
This adds support for GitHub App authentication using the main class which is simpler for users (see #1855)

```python
import Github

github_client = Github(app_id=MY_APP_ID, app_private_key=MY_PRIVATE_KEY)
```

Also, the credentials are refreshed when needed (tokens lasts for one hour). This should happen transparently to the users, as the token expiration is checked at every call.

Caveats and other notes:
- Uses the first found installation for an app. This means that it only works in the context of private apps
- I'm not sure how to add tests correctly and there does not seem to be any for `GithubIntegration`, so I skipped that part. Manual testing works and existing tests are not broken
- I could not resist fixing some type stubs along the way

### CODE REVIEW

File: github/Consts.py
Comment (dblanchette) [inline github/Consts.py:130]: Moved those here since they are used in 2 files now (`MainClass` and  `GithubIntegration`)

File: github/GithubException.py
Comment (dblanchette) [inline github/GithubException.py:41]: I had a 404 while testing and had an exception because the headers parameters was not passed.

File: github/GithubIntegration.py
Comment (dblanchette) [inline github/GithubIntegration.py:1]: Moved from `MainClass`, it's already big as it is and that does not really belong there.

File: github/GithubIntegration.py
Comment (dblanchette) [inline github/GithubIntegration.py:88]: Let me know if that does not suit you. I feel that this call is badly named

File: github/GithubIntegration.py
Comment (dblanchette) [inline github/GithubIntegration.py:112]: This is the part I added in that file. Users having a private app with more than one installations can use this call as the main class always uses the first found installation.

File: github/Requester.py
Comment (dblanchette) [inline github/Requester.py:313]: The heart of the changes I made is here!

**Review by dblanchette — COMMENTED** [review]
Added some comments for easier reviewing.

**Comment by maxdanilov** [comment]
also very interested in seeing this implemented, are there any updates on this feature?

**Comment by simkimsia** [comment]
Will this be integrated?

**Comment by vitalii94lukianenko** [comment]
Hi, Could you merge this PR?

**Comment by ThomasJRyan** [comment]
Would definitely appreciate this feature being merged in

File: github/MainClass.pyi
Comment (EnricoMi) [inline github/MainClass.pyi:41]: Should be ` = ...,` to not repeat the default value (see all other arguments in that signature).

Comment (EnricoMi) [reply github/MainClass.pyi:41]: Same in `Requester.pyi`.

File: github/GithubException.py
Comment (EnricoMi) [reply github/GithubException.py:41]: I think fixing the calling code as in #2079 is better than changing the signature.

File: github/GithubIntegration.py
Comment (ThomasJRyan) [inline github/GithubIntegration.py:125]: If I'm not mistaken, this line makes the `base_url` passed in through the `__init__` useless. Makes it impossible to use it with Github Enterprise
```suggestion
                base_url=self.base_url,
```

**Comment by dblanchette** [comment]
Thanks for the review! I made the requested changes.

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py]: Good catch!

**Comment by dblanchette** [comment]
@EnricoMi @ThomasJRyan Anything else I should do?

File: github/ApplicationOAuth.pyi
Comment (EnricoMi) [inline github/ApplicationOAuth.pyi:16]: `.pyi` files usually do not have default values but `...` instead, default values are defined in the respective `.py` files, same issue in other files

File: github/Requester.py
Comment (EnricoMi) [inline github/Requester.py:363]: Should the token be refreshed some time before `expires_at` to make sure that after calling this method the request does not exceed `expired_at`? Or will e request with an expired token be retried anyway?

Comment (dblanchette) [reply github/Requester.py:363]: I put a 20 seconds delay (see `ACCESS_TOKEN_REFRESH_THRESHOLD_SECONDS`) to make sure there is still enough time to make the request.

File: github/ApplicationOAuth.pyi
Comment (dblanchette) [reply github/ApplicationOAuth.pyi]: Sorry about those, I had never used pyi files before, so I'm used to adding the hints where the default value is. Should all be fixed now.

File: github/Requester.py
Comment (EnricoMi) [reply github/Requester.py:363]: right, I misread the logic.

File: github/CheckSuite.pyi
Comment (EnricoMi) [inline github/CheckSuite.pyi:52]: here are still some more default values in a `.pyi` file

File: github/Commit.pyi
Comment (EnricoMi) [inline github/Commit.pyi:49]: and here

**Comment by dblanchette** [comment]
I went and look at all the files I changed to make sure there were no default values left in pyi. Crossing my fingers that everything is ok now 🤞

**Comment by dblanchette** [comment]
I hate to ask again, but is it all good?

I'm looking forward to go back to using the official library instead of our fork.

Thanks for your work maintaining this library!

**Comment by aabouzaid** [comment]
Any plans to merge this soon, please?

**Comment by jonapich** [comment]
The last merged PR is from December 20th 2021 (3 months ago) and now there's over 60 pull requests waiting. It feels like there's a need for additional maintainers?

**Comment by aaronn** [comment]
I can't seem to get my .pem file loaded into the `app_private_key=MY_PRIVATE_KEY` field with `open(file).read` nor `just the filename. Is there something I'm missing?

**Comment by dblanchette** [comment]
@aaronn 
> I can't seem to get my .pem file loaded into the `app_private_key=MY_PRIVATE_KEY` field with `open(file).read()` nor just the filename. Is there something I'm missing?

I'm not really sure what's wrong. It works for us with `MY_PRIVATE_KEY` being the private key as a decoded string beginning with `-----BEGIN RSA PRIVATE KEY-----` and ending with `-----END RSA PRIVATE KEY-----` (included). The line feeds are also included and they are `\n` (not `\r\n` i.e. on Windows). Maybe you need to set an encoding when opening the file? Otherwise, I'd try with the rest API directly and make sure that works.

Hope that helps!

**Comment by yskyba-kenshoo** [comment]
Hi, any updates here? Could you merge it?

**Comment by alelapi** [comment]
I think there is no more need for this PR. 
GitHub App authentication can be implemented so:

```
integration = GithubIntegration(app_id, private_key)
installation_id = integration.get_installation(org, repo_name).id
access_token = integration.get_access_token(installation_id).token
github_client =  Github(login_or_token=self.access_token)
```

**Comment by jnewland** [comment]
> I think there is no more need for this PR.

I'm using this PR to support [creating a GitHub App from a manifest](https://docs.github.com/en/rest/apps/apps#create-a-github-app-from-a-manifest), which expects the integrator to handle the case where an app has been installed on an account or organization but not yet installed on any repositories.

**Comment by skeggse** [comment]
> I think there is no more need for this PR. GitHub App authentication can be implemented so:
> 
> ```
> integration = GithubIntegration(app_id, private_key)
> installation_id = integration.get_installation(org, repo_name).id
> access_token = integration.get_access_token(installation_id).token
> github_client =  Github(login_or_token=self.access_token)
> ```

That works fine for short-lived interactions, but if you want to interact with the Github API for hours or days, that access token will expire. It's better to have this implemented in PyGithub itself so it can handle automatic token refreshes.

**Comment by alelapi** [comment]
> 

Only for my clarity's sake, could you provide an example of long interaction that cannot be replaced by using webhooks? 
GitHub forces the token's expiration time to 600 s for some reason I think

**Comment by alelapi** [comment]
> > I think there is no more need for this PR. GitHub App authentication can be implemented so:
> > ```
> > integration = GithubIntegration(app_id, private_key)
> > installation_id = integration.get_installation(org, repo_name).id
> > access_token = integration.get_access_token(installation_id).token
> > github_client =  Github(login_or_token=self.access_token)
> > ```
> 
> That works fine for short-lived interactions, but if you want to interact with the Github API for hours or days, that access token will expire. It's better to have this implemented in PyGithub itself so it can handle automatic token refreshes.

Only for my clarity's sake, could you provide an example of long interaction that cannot be replaced by using webhooks? 
GitHub forces the token's expiration time to 600 s for some reason I think

**Comment by cmac4603** [comment]
> Uses the first found installation for an app. This means that it only works in the context of private apps

So if you have an app installed in two organizations GitHubIntegration will no longer work? This would be a real issue for myself and potentially others I would think

**Comment by skwashd** [comment]
I have a need for this. I am happy to assist with making changes needed to get it merged.

@dblanchette do you want some help with adding tests? I am also able to assist with fixing "Uses the first found installation for an app. This means that it only works in the context of private apps" so it works for multiple installations.

@s-t-e-v-e-n-k is there anything else you want to see here?

**Comment by sfdye** [comment]
Looks like there are conflicts that need to be manually resolved on this PR

**Comment by dblanchette** [comment]
> @dblanchette do you want some help with adding tests? I am also able to assist with fixing "Uses the first found installation for an app. This means that it only works in the context of private apps" so it works for multiple installations.

Of course, help would be appreciated. Feel free to open a PR in the coveord fork so that I can add it gets added to this one or build a new PR on top of this one.

> Looks like there are conflicts that need to be manually resolved on this PR

All resolved

**Comment by ammarmallik** [comment]
> Of course, help would be appreciated. Feel free to open a PR in the coveord fork so that I can add it gets added to this one or build a new PR on top of this one.

@dblanchette here's the [PR](https://github.com/coveord/PyGithub/pull/28) that includes changes related to test cases and some new features related to fetching installation for user/repo/org. Also regarding multiple installations, currently Github doesn't have an API/functionality to generate the token that can cover multiple installations. Currrently, you can have one token per installation. As long as you select the right installation ID, you are good to go.

I've requested feedback from the community to check if there's any chance of having multiple installations feature provided by Github for Github Apps. I'll share their feedback here once I have it.

CC: @skwashd

**Comment by dblanchette** [comment]
@ammarmallik I merged your changes https://github.com/coveord/PyGithub/pull/28 and they are now integratred with this PR.

I agree with requesting the installation id, even if it is one more call for our use case. Taking the first installation id could result in surprising behavior!

@sfdye Could you take a look please?

File: github/Requester.py
Comment (rcoup) [inline github/Requester.py:407]: This isn't thread-safe at the moment, and race conditions are quite likely to occur given the nature of the expiry. A few ideas:

1. an `RLock` around refreshing
2. put `_install_authorization` & `_authorization_header` in a thread-local, but that's more calls to the GH API even when there's no race.
3. do something like this using a tuple/namedtuple so if there is a refresh-race the instance attribute is assigned atomically. And change other uses of `__installation_authorization`/`__authorizationHeader` to reference the tuple.
```
install_authorization = self._get_installation_authorization()
authorization_header = f"token {self.__installation_authorization.token}"
# atomic assignment
self.__authorization = (install_authorization, authorization_header)
```

**Comment by skwashd** [comment]
@sfdye can this be merged as is or does the issue flagged by @rcoup need to be resolved first?

**Review by rcoup — COMMENTED** [review]
I'm happy with the credentials refresh, but I haven't reviewed the rest of the PR in depth. Beyond using it :-)

**Comment by dblanchette** [comment]
@rcoup Would you be the one to review it?

We've been using this in production since I posted this PR in June 2021 and haven't got issues.

**Comment by rcoup** [comment]
@dblanchette I'm not involved in PyGithub beyond being a user. We're using this in an internal application as well without issues.

I think @s-t-e-v-e-n-k & @sfdye are the active contributors.

**Comment by s-t-e-v-e-n-k** [comment]
There are a large amount of unrelated changes in this PR, would you mind rebasing it?

**Comment by dblanchette** [comment]
@s-t-e-v-e-n-k The unrelated changes are fixes around .pyi files not having the same values as the .py.

I removed them in my most recent commit so that the changes in the PR are only the ones related to GitHub App authentication. I can submit them in another PR after this one is merged.

File: github/GithubIntegration.py
Comment (s-t-e-v-e-n-k) [inline github/GithubIntegration.py:131]: Why are you catching and then raising these two exceptions here?

File: github/GithubIntegration.py
Comment (s-t-e-v-e-n-k) [reply github/GithubIntegration.py:143]: This is the existing call? It should probably also state it's deprecated in the docstring. Also, it should mention the class, not just the bare function.

File: github/GithubIntegration.pyi
Comment (s-t-e-v-e-n-k) [inline github/GithubIntegration.pyi:16]: dict of what?

File: github/MainClass.py
Comment (s-t-e-v-e-n-k) [inline github/MainClass.py:105]: `app_token_permissions` in the signature, and `app_token_scopes` here.

File: github/MainClass.py
Comment (s-t-e-v-e-n-k) [inline github/MainClass.py:789]: Now for my big question -- is this an API break?

File: github/MainClass.pyi
Comment (s-t-e-v-e-n-k) [inline github/MainClass.pyi:36]: dict containing what types?

File: github/MainClass.pyi
Comment (s-t-e-v-e-n-k) [inline github/MainClass.pyi:91]: This is an unrelated change, but also a welcome one.

File: github/Requester.pyi
Comment (s-t-e-v-e-n-k) [inline github/Requester.pyi:63]: There are still unrelated changes in this file, but I understand it's hard to untangle them.

File: github/SelfHostedActionsRunner.pyi
Comment (s-t-e-v-e-n-k) [inline github/SelfHostedActionsRunner.pyi:1]: This looks unrelated?

**Review by s-t-e-v-e-n-k — CHANGES_REQUESTED** [review]
Broadly, this is looking excellent, albeit a little terrifying. I still have some small concerns, which I've marked inline.

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py:131]: I honestly don't remember, but I see other code that does this, maybe it is to alert readers as to the exceptions that can be raised by this method?

This could also be achieved by using `:raises UnknownObjectException:` and `:raises GithubException:` in the docstring. Do you prefer I change it?

File: github/MainClass.pyi
Comment (dblanchette) [reply github/MainClass.pyi:36]: I'll change those to `Dict[str, str]`. The GitHub reference just states `object`

File: github/SelfHostedActionsRunner.pyi
Comment (dblanchette) [reply github/SelfHostedActionsRunner.pyi:1]: Missed that one, sorry. We added a bunch of stuff for self-hosted runners in our fork.

Comment (dblanchette) [reply github/SelfHostedActionsRunner.pyi:1]: Actually, this is a change caused by wrong line endings. By loading the file, my IDE changed it from `\r\n` to `\n`. Since the latter is used all over the repo, I think this is a good change. You can see there is no actual change by checking `Hide whitespaces` in the cog wheel menu at the top of the PR.

File: github/GithubIntegration.py
Comment (s-t-e-v-e-n-k) [reply github/GithubIntegration.py:131]: We commonly catch the exceptions to change the return type -- for example `Team.get_repo_permission`, so I don't see any value for us catching the exceptions and then re-raising them. We may want to document in the docstrings which exceptions can be raised, but I'd prefer we punt that to the future.

File: github/MainClass.py
Comment (s-t-e-v-e-n-k) [reply github/MainClass.py:789]: Perhaps with a deprecation message as well? Then we don't have to ship it forever.

File: github/MainClass.py
Comment (EnricoMi) [inline github/MainClass.py:89]: Looks like these four go together in many places in this PR. What do you think about moving them into a class so we are carrying around only a single argument, e.g. `app_auth`.

Ideally, the now four existing ways for authentication (token, login and password, jwt, app auth) should be a single argument `auth` of a single base class (e.g. `GithubAuth`) with different implementations (`GitHubTokenAuth`, `GitHubLoginAuth`, `GitHubJwtAuth`, `GitHubAppAuth`). That would simplify a lot of signatures and stop growing signatures when new auths are added. But that is out-of-scope of this PR.

File: github/MainClass.py
Comment (dblanchette) [reply github/MainClass.py:789]: I'm not sure how to add a message when a user imports from the wrong location, do you have something in mind?

File: github/MainClass.py
Comment (dblanchette) [reply github/MainClass.py]: That is a good point and it's done :)

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py]: I removed the useless exception handling.

**Comment by s-t-e-v-e-n-k** [comment]
Now fails tests with ` E       AttributeError: module 'github' has no attribute 'AppAuthentication'`

File: github/GithubIntegration.py
Comment (EnricoMi) [inline github/GithubIntegration.py:90]: The

    valid for 60 seconds by default. Could be set to a maximum of 600 seconds.

does not make sense in this method any more, as the method does not have the `expiry` parameter. The comment should refer to `self.jwt_expiry.

This comment should go into the `__init__` to document parameter `jwt_expiry`.

File: github/GithubIntegration.py
Comment (EnricoMi) [inline github/GithubIntegration.py:93]: GitHub documents `iat` as

    To protect against clock drift, we recommend you set this 60 seconds in the past.

That should be another parameter to this class, which defaults to 60s.

File: github/Consts.py
Comment (EnricoMi) [reply github/Consts.py:135]: Comparing this is really difficult. Would have been better to do the refactoring (without changes) in a separate PR and base this PR on that. Then we would see only changes, not moving unchanged code around.

File: github/AppAuthentication.py
Comment (EnricoMi) [inline github/AppAuthentication.py:32]:
```suggestion
        assert isinstance(app_id, (int, str)), app_id
        assert isinstance(private_key, str)
        assert isinstance(installation_id, int), installation_id
        assert isinstance(token_permissions, int), token_permissions
        self.app_id = app_id
```

File: github/GithubIntegration.py
Comment (EnricoMi) [inline github/GithubIntegration.py:33]: Value of parameter `private_key` should not be leaked, even if it is not a string.
```suggestion
        assert isinstance(private_key, str)
```

File: github/Requester.py
Comment (EnricoMi) [inline github/Requester.py:396]: It does not look like this method really requires `self.__app_auth is not None`, it is sufficient that `_get_installation_authorization` has that assertion. This method should not care, it is internal to `_get_installation_authorization`.

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py]: Replaced by a message not containing the private key

**Comment by dblanchette** [comment]
> `Now fails tests with  E       AttributeError: module 'github' has no attribute 'AppAuthentication'`

That is fixed. I must say I find the imports in this project to be done in a peculiar fashion.

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py]: I moved them to `Consts` where they are documented (link to the GitHub online documentation)

File: github/Consts.py
Comment (dblanchette) [reply github/Consts.py]: Changed

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py]: Moved

File: github/GithubIntegration.py
Comment (dblanchette) [reply github/GithubIntegration.py:100]: I added a parameter. I don't think it will be used since users will probably not need to set that to a different value, but it does not hurt to have the option!

File: github/Requester.py
Comment (dblanchette) [reply github/Requester.py]: Indeed, removed

**Comment by dblanchette** [comment]
> Comparing this is really difficult. Would have been better to do the refactoring (without changes) in a separate PR and base this PR on that. Then we would see only changes, not moving unchanged code around.

Agreed. I tend to go all out in refactoring as you've seen with the unrelated changes. I'll know to do that in the future.

Thank you both for the reviews!

File: github/AppAuthentication.py
Comment (ammarmallik) [reply github/AppAuthentication.py:37]: Yes, they are optional and this is the correct type: `Optional[Dict[str, str]]`

**Comment by dblanchette** [comment]
All good now? I'm excited to see this release :)

**Review by s-t-e-v-e-n-k — APPROVED** [review]

**Comment by EnricoMi** [comment]
@dblanchette can you please take a look at #2420 /  #2425?

**Comment by EnricoMi** [comment]
@dblanchette another issue likely being introduced by this PR: #2431

**Comment by EnricoMi** [comment]
Another issue: #2432

**Comment by EnricoMi** [comment]
This is quite some fallout, two more issues reported: #2436, #2441.

**Pull request merged**
